### ISSUE TITLE
write_basic_package_version_file COMPATIBILITY should be AnyNewerVersion, not SameMajorVersion

### ISSUE THREAD

**Issue by kkofler** [scipopt/scip#116]
See scipopt/JSCIPOpt#52 and in particular https://github.com/scipopt/JSCIPOpt/issues/52#issuecomment-2423333593

> This is actually a misfeature of the SCIP build system.
> 
> `scip/src/CMakeLists.txt` has this:
> 
> ```cmake
> write_basic_package_version_file(
> ${PROJECT_BINARY_DIR}/scip-config-version.cmake
>     VERSION ${SCIP_VERSION_MAJOR}.${SCIP_VERSION_MINOR}.${SCIP_VERSION_PATCH}
>     COMPATIBILITY SameMajorVersion
>     )
> ```
> 
> The `SameMajorVersion` there is causing this behavior. See the generated `scip-config-version.cmake` to see the behavior. (You cannot even specify a range like `8.0.0...9.9.9` or `8.0.0...<10.0.0`, it will reject any range that is not restricted to a single major version.)
> 
> I will work around this as described in my previous comment, but it should really be fixed in SCIP. The compatibility should be `AnyNewerVersion`. If a dependent package really wants to specify SCIP 8.x.x only, it should request the range `8.0.0...<9.0.0` explicity. It should not be done by default, especially because there is no way to opt out of this behavior.

**Comment by kkofler** [comment]
`SameMajorVersion` is intended for packages like Qt where code written for Qt 5 is very unlikely to compile unmodified with Qt 6. Not for packages like SCIP that are almost completely backwards-compatible at source level (and even binary compatible if you do not use the macros for the getters).

**Comment by svigerske** [comment]
We are breaking the API between major versions of SCIP, but yeah, there are attempts to not do that too much or give time to migrate (by keeping old functions as deprecated before removing). We'll consider replacing it with `AnyNewerVersion`.

**Comment by svigerske** [comment]
Looks like this isn't happening. "almost completely backwards-compatible" wasn't considered sufficient to use `AnyNewerVersion`. The [cmake docu](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#generating-a-package-version-file) also says "This mode should be used for packages which are fully backward compatible, also across major versions." With SCIP's rather large API, there is usually some backward incompatible change in a major release (e.g., "Deleted and changed API methods" in https://www.scipopt.org/doc-9.1.1/html/RN9.php#RN900), even though it may be in functions that aren't used by many.

**Comment by kkofler** [comment]
(And by the way, my pedantic mathematician mind says that you should close this as "Not Planned" rather than "Completed" since you do not plan to implement the requested change.)

**Comment by svigerske** [comment]
I understand that the situation is not ideal. But if we change it to `AnyNewerVersion`, then anyone who looks this up in the cmake docu will think that SCIP's API is completely backward compatible among major versions, not just "almost". We will then soon have someone who requests to change it back to `SameMajorVersion`.

Even if we would ensure that the API is backward compatible for one major release, we wouldn't want to promise this for "any newer versions". It's also difficult to say from the SCIP point of view whether the API changes are "harmless" to SCIP users or require action. It is rather up to the maintainer of a package like JSCIPOpt to decide which SCIP versions they are compatible with. That you cannot specify this within `find_package` is a deficiency `find_package` I would say.

**Comment by kkofler** [comment]
> But if we change it to `AnyNewerVersion`, then anyone who looks this up in the cmake docu will think that SCIP's API is completely backward compatible among major versions, not just "almost".

Why would anyone who is not a developer of SCIP (and those who are already know what the API compatibility really looks like) look up the documentation of that flag to `write_basic_package_version_file` in the CMake documentation? This is a source line in the SCIP build system. It is not installed. Only the generated output is, which does not spell out what flags it is generated with.

I had to actually search for `scip-config-version.cmake` in the SCIP source code to find this line, after already figuring out through debugging that `scip-config-version.cmake` was where the version was rejected, and I only ended up doing this because `find_package` did not work the way I expected it to, which is to accept any newer version (which used to be how `find_package` always worked before `write_basic_package_version_file` was introduced, I was not even aware that packages' config modules can now override this). There is no way that a SCIP user / application developer will just out of the blue (without encountering such an error, which they will not with `AnyNewerVersion`) look at that line to figure out what the SCIP API promise actually is.

And in case someone really does, a comment above that line could very easily set that misconception straight.
